File: src/lib/crc.ts
import { invoke } from "@tauri-apps/api/core";

import type {
  CrcDraft,
  CrcInputFormat,
} from "@/features/crc/crc-config";

export type CrcCalculationRequest = {
  payload: string;
  inputFormat: CrcInputFormat;
  width: number;
  polynomial: string;
  init: string;
  xorOut: string;
  reflectIn: boolean;
  reflectOut: boolean;
};

export type CrcCalculationResult = {
  value: string;
  decimal: string;
  binary: string;
  bytes: number;
  width: number;
  mask: string;
  normalized: {
    polynomial: string;
    init: string;
    xorOut: string;
    reflectedPolynomial: string;
  };
};

export async function calculateCrc(
  payload: string,
  inputFormat: CrcInputFormat,
  draft: CrcDraft,
): Promise<CrcCalculationResult> {
  return invoke<CrcCalculationResult>("calculate_crc", {
    request: {
      payload,
      inputFormat,
      width: Number(draft.width),
      polynomial: draft.polynomial,
      init: draft.init,
      xorOut: draft.xorOut,
      reflectIn: draft.reflectIn,
      reflectOut: draft.reflectOut,
    } satisfies CrcCalculationRequest,
  });
}

File: src-tauri/src/domain/crc.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CrcCalculationRequest {
    pub payload: String,
    pub input_format: String,
    pub width: u32,
    pub polynomial: String,
    pub init: String,
    pub xor_out: String,
    pub reflect_in: bool,
    pub reflect_out: bool,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CrcCalculationResult {
    pub value: String,
    pub decimal: String,
    pub binary: String,
    pub bytes: usize,
    pub width: u32,
    pub mask: String,
    pub normalized: CrcNormalizedParameters,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CrcNormalizedParameters {
    pub polynomial: String,
    pub init: String,
    pub xor_out: String,
    pub reflected_polynomial: String,
}